File: client_windows_native/Services/StreamingASRClient.cs
using System;
using System.Collections.Generic;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using VoiceTyper.Core;
using VoiceTyper.Support;

namespace VoiceTyper.Services;

/// <summary>
/// WebSocket 流式语音识别客户端。一次录音对应一个实例，不复用。
/// 生命周期：Connect → (SendAudio × N → Finalize) → [OnFinal/OnError] → Close。
///
/// 所有事件回调通过 <see cref="UiDispatcher"/> 投递回 UI 线程。
/// </summary>
internal sealed class StreamingASRClient : IDisposable
{
    public Action<string>? OnPartial;
    public Action<string>? OnFinal;
    public Action<string>? OnError;

    private ClientWebSocket? _ws;
    private CancellationTokenSource? _cts;
    private Task? _receiveTask;
    private readonly SemaphoreSlim _sendLock = new(1, 1);
    private bool _closed;
    private bool _finalReceived;

    public async Task ConnectAsync(ServerConfig server, IReadOnlyList<string> hotwords, bool llmRecorrect, CancellationToken ct = default)
    {
        if (_ws is not null) throw new InvalidOperationException("StreamingASRClient 已连接");

        var scheme = "ws";
        var url = $"{scheme}://{server.Host}:{server.Port}/recognize/stream?llm_recorrect={(llmRecorrect ? "true" : "false")}";

        _ws = new ClientWebSocket();
        var trimmedKey = (server.ApiKey ?? "").Trim();
        if (trimmedKey.Length > 0)
        {
            _ws.Options.SetRequestHeader("Authorization", $"Bearer {trimmedKey}");
        }

        _cts = new CancellationTokenSource();
        var linkedCt = CancellationTokenSource.CreateLinkedTokenSource(_cts.Token, ct).Token;

        try
        {
            await _ws.ConnectAsync(new Uri(url), linkedCt).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            AppLog.Error("network", "WS 连接失败", ex);
            try { _ws.Dispose(); } catch { }
            _ws = null;
            _cts.Dispose();
            _cts = null;
            throw;
        }

        // 发送 start 帧
        var startPayload = new Dictionary<string, object?>
        {
            ["type"] = "start",
            ["hotwords"] = string.Join(" ", hotwords),
            ["sample_rate"] = AppConstants.TargetSampleRate,
        };
        await SendJsonAsync(startPayload, linkedCt).ConfigureAwait(false);

        // 启动接收循环
        _receiveTask = Task.Run(() => ReceiveLoopAsync(_cts.Token));
    }

    public void SendAudio(byte[] data)
    {
        if (_closed || _ws is null) return;
        _ = SendAudioAsync(data);
    }

    public void FinalizeStream()
    {
        if (_closed || _ws is null) return;
        _ = SendJsonAsync(new Dictionary<string, object?> { ["type"] = "finalize" }, CancellationToken.None);
    }

    public void Close()
    {
        if (_closed) return;
        _closed = true;

        try { _cts?.Cancel(); } catch { }
        var ws = _ws;
        _ws = null;

        if (ws is not null)
        {
            // 异步、最佳努力关闭；不等待
            _ = Task.Run(async () =>
            {
                try
                {
                    if (ws.State == WebSocketState.Open || ws.State == WebSocketState.CloseReceived)
                    {
                        using var timeoutCts = new CancellationTokenSource(TimeSpan.FromSeconds(1));
                        await ws.CloseAsync(WebSocketCloseStatus.NormalClosure, null, timeoutCts.Token).ConfigureAwait(false);
                    }
                }
                catch { /* swallow */ }
                finally
                {
                    try { ws.Dispose(); } catch { }
                }
            });
        }

        try { _cts?.Dispose(); } catch { }
        _cts = null;
    }

    public void Dispose() => Close();

    private async Task SendAudioAsync(byte[] data)
    {
        var ws = _ws;
        if (ws is null || _closed) return;

        await _sendLock.WaitAsync().ConfigureAwait(false);
        try
        {
            if (_closed) return;
            if (ws.State != WebSocketState.Open) return;
            await ws.SendAsync(new ArraySegment<byte>(data), WebSocketMessageType.Binary, endOfMessage: true, CancellationToken.None)
                .ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            if (!_closed)
            {
                AppLog.Warn("network", $"音频帧发送失败: {ex.Message}");
            }
        }
        finally
        {
            _sendLock.Release();
        }
    }

    private async Task SendJsonAsync(Dictionary<string, object?> payload, CancellationToken ct)
    {
        var ws = _ws;
        if (ws is null || _closed) return;

        var json = JsonSerializer.Serialize(payload);
        var bytes = Encoding.UTF8.GetBytes(json);

        await _sendLock.WaitAsync(ct).ConfigureAwait(false);
        try
        {
            if (_closed) return;
            if (ws.State != WebSocketState.Open) return;
            await ws.SendAsync(new ArraySegment<byte>(bytes), WebSocketMessageType.Text, endOfMessage: true, ct).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            if (!_closed)
            {
                AppLog.Warn("network", $"控制帧发送失败: {ex.Message}");
            }
        }
        finally
        {
            _sendLock.Release();
        }
    }

    private async Task ReceiveLoopAsync(CancellationToken ct)
    {
        var ws = _ws;
        if (ws is null) return;

        var buffer = new byte[16 * 1024];
        var sb = new StringBuilder();

        try
        {
            while (!ct.IsCancellationRequested && ws.State == WebSocketState.Open)
            {
                sb.Clear();
                WebSocketReceiveResult result;
                do
                {
                    result = await ws.ReceiveAsync(new ArraySegment<byte>(buffer), ct).ConfigureAwait(false);
                    if (result.MessageType == WebSocketMessageType.Close)
                    {
                        return;
                    }
                    sb.Append(Encoding.UTF8.GetString(buffer, 0, result.Count));
                } while (!result.EndOfMessage);

                HandleMessage(sb.ToString());
            }
        }
        catch (OperationCanceledException)
        {
            // 正常关闭
        }
        catch (Exception ex)
        {
            if (!_closed && !_finalReceived)
            {
                AppLog.Warn("network", $"WS receive 失败: {ex.Message}");
                UiDispatcher.Post(() => OnError?.Invoke("连接中断"));
            }
        }
    }

    private void HandleMessage(string raw)
    {
        ServerMessage msg;
        try
        {
            msg = ServerMessage.Parse(raw);
        }
        catch (Exception ex)
        {
            AppLog.Warn("network", $"解析服务端消息失败: {ex.Message}");
            return;
        }

        switch (msg.Type)
        {
            case ServerMessageType.Partial:
                if (!string.IsNullOrEmpty(msg.Text))
                {
                    UiDispatcher.Post(() => OnPartial?.Invoke(msg.Text!));
                }
                break;
            case ServerMessageType.Final:
                _finalReceived = true;
                AppLog.Info("network", $"final: text={Truncate(msg.Text ?? "", 80)} asr={msg.AsrElapsed} llm={msg.LlmElapsed}");
                UiDispatcher.Post(() => OnFinal?.Invoke(msg.Text ?? ""));
                break;
            case ServerMessageType.Error:
                AppLog.Error("network", $"服务端错误: {msg.Code} {msg.Message}");
                UiDispatcher.Post(() => OnError?.Invoke(msg.Message ?? "服务端错误"));
                break;
        }
    }

    private static string Truncate(string s, int max) =>
        s.Length <= max ? s : s.Substring(0, max) + "...";

    // ─── 服务端消息 ─────────────────────────────────────────────────

    private enum ServerMessageType { Unknown, Partial, Final, Error }

    private readonly struct ServerMessage
    {
        public ServerMessageType Type { get; init; }
        public string? Text { get; init; }
        public int Seq { get; init; }
        public double? AsrElapsed { get; init; }
        public double? LlmElapsed { get; init; }
        public string? Code { get; init; }
        public string? Message { get; init; }

        public static ServerMessage Parse(string raw)
        {
            using var doc = JsonDocument.Parse(raw);
            var root = doc.RootElement;
            if (!root.TryGetProperty("type", out var typeProp))
            {
                return new ServerMessage { Type = ServerMessageType.Unknown };
            }
            var type = typeProp.GetString();

            return type switch
            {
                "partial" => new ServerMessage
                {
                    Type = ServerMessageType.Partial,
                    Text = root.TryGetProperty("text", out var t) ? t.GetString() : "",
                    Seq = root.TryGetProperty("seq", out var s) && s.TryGetInt32(out var sv) ? sv : 0,
                },
                "final" => new ServerMessage
                {
                    Type = ServerMessageType.Final,
                    Text = root.TryGetProperty("text", out var t) ? t.GetString() : "",
                    AsrElapsed = root.TryGetProperty("asrElapsed", out var a) && a.TryGetDouble(out var av) ? av : null,
                    LlmElapsed = root.TryGetProperty("llmElapsed", out var l) && l.TryGetDouble(out var lv) ? lv : null,
                },
                "error" => new ServerMessage
                {
                    Type = ServerMessageType.Error,
                    Code = root.TryGetProperty("code", out var c) ? c.GetString() : "unknown",
                    Message = root.TryGetProperty("message", out var m) ? m.GetString() : "",
                },
                _ => new ServerMessage { Type = ServerMessageType.Unknown },
            };
        }
    }
}

File: client_macos_swift/Sources/VoiceTyperApp/Services/StreamingASRClient.swift
import Foundation

// MARK: - 服务端消息结构

private enum ServerMessage {
    case partial(text: String, seq: Int)
    case final(text: String, asrElapsed: Double?, llmElapsed: Double?)
    case warning(code: String, message: String)
    case error(code: String, message: String)
    case unknown
}

private func parseServerMessage(_ raw: String) -> ServerMessage {
    guard let data = raw.data(using: .utf8),
          let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let type = json["type"] as? String
    else {
        return .unknown
    }

    switch type {
    case "partial":
        let text = json["text"] as? String ?? ""
        let seq = json["seq"] as? Int ?? 0
        return .partial(text: text, seq: seq)

    case "final":
        let text = json["text"] as? String ?? ""
        let asrElapsed = json["asrElapsed"] as? Double
        let llmElapsed = json["llmElapsed"] as? Double
        return .final(text: text, asrElapsed: asrElapsed, llmElapsed: llmElapsed)

    case "warning":
        let code = json["code"] as? String ?? "unknown"
        let message = json["message"] as? String ?? ""
        return .warning(code: code, message: message)

    case "error":
        let code = json["code"] as? String ?? "unknown"
        let message = json["message"] as? String ?? ""
        return .error(code: code, message: message)

    default:
        return .unknown
    }
}

// MARK: - StreamingASRClient

/// WebSocket 客户端，管理单次录音会话的全双工流式通信。
///
/// 生命周期：connect → (sendAudio × N → finalize) → [onFinal/onError 触发] → close
/// 每次录音新建一个实例，不复用。
@MainActor
final class StreamingASRClient {
    var onPartial: ((String) -> Void)?
    var onFinal: ((String) -> Void)?
    /// 非致命提示（如 feed_failed）。连接仍存活，finalize 可继续。
    var onWarning: ((String) -> Void)?
    var onError: ((String) -> Void)?

    private var task: URLSessionWebSocketTask?
    private var receiveLoopTask: Task<Void, Never>?
    /// finalize 后等待 final 帧的超时任务；收到 final/error 或 close 时取消。
    private var finalizeTimeoutTask: Task<Void, Never>?
    private var closed = false

    // MARK: - 连接

    func connect(server: ServerConfig, hotwords: [String], llmRecorrect: Bool) throws {
        var components = URLComponents()
        components.scheme = server.wsScheme
        components.host = server.host
        components.port = server.port
        components.path = "/recognize/stream"
        components.queryItems = [
            URLQueryItem(name: "llm_recorrect", value: llmRecorrect ? "true" : "false"),
        ]

        guard let url = components.url else {
            throw URLError(.badURL)
        }

        var request = URLRequest(url: url)
        request.timeoutInterval = 5.0
        let trimmedKey = server.apiKey.trimmingCharacters(in: .whitespacesAndNewlines)
        if !trimmedKey.isEmpty {
            request.setValue("Bearer \(trimmedKey)", forHTTPHeaderField: "Authorization")
        }

        let wsTask = URLSession.shared.webSocketTask(with: request)
        self.task = wsTask
        wsTask.resume()

        // 发送 start 帧
        let startPayload: [String: Any] = [
            "type": "start",
            "hotwords": hotwords.joined(separator: " "),
            "sample_rate": 16000,
        ]
        sendJSON(startPayload)

        // 启动接收循环
        receiveLoopTask = Task { [weak self] in
            await self?.receiveLoop()
        }
    }

    // MARK: - 发送

    func sendAudio(_ data: Data) {
        guard !closed else { return }
        task?.send(.data(data)) { error in
            if let error {
                AppLog.network.error("音频帧发送失败: \(error.localizedDescription, privacy: .public)")
            }
        }
    }

    /// 请求离线复识别的最终结果。
    /// - Parameter timeout: 等待 `final` 帧的最长时间；超时视为识别失败，触发 onError，
    ///   避免服务端卡住 / 网络半开时 HUD 永久停在"识别中"。传 <= 0 表示不设超时。
    func finalize(timeout: TimeInterval) {
        guard !closed else { return }
        sendJSON(["type": "finalize"])

        finalizeTimeoutTask?.cancel()
        guard timeout > 0 else { return }
        finalizeTimeoutTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
            guard !Task.isCancelled, let self, !self.closed else { return }
            AppLog.network.error("finalize 超时（\(timeout, privacy: .public)s），未收到 final")
            self.onError?("识别超时")
        }
    }

    func close() {
        guard !closed else { return }
        closed = true
        finalizeTimeoutTask?.cancel()
        finalizeTimeoutTask = nil
        receiveLoopTask?.cancel()
        receiveLoopTask = nil
        task?.cancel(with: .normalClosure, reason: nil)
        task = nil
    }

    // MARK: - 接收循环

    private func receiveLoop() async {
        guard let task else { return }

        while !Task.isCancelled {
            do {
                let message = try await task.receive()
                switch message {
                case .string(let raw):
                    handleServerMessage(raw)
                case .data(let data):
                    if let raw = String(data: data, encoding: .utf8) {
                        handleServerMessage(raw)
                    }
                @unknown default:
                    break
                }
            } catch {
                if !closed && !Task.isCancelled {
                    AppLog.network.error("WS receive 失败: \(error.localizedDescription, privacy: .public)")
                    onError?("连接中断")
                }
                break
            }
        }
    }

    @MainActor
    private func handleServerMessage(_ raw: String) {
        switch parseServerMessage(raw) {
        case .partial(let text, _):
            if !text.isEmpty {
                onPartial?(text)
            }
        case .final(let text, let asr, let llm):
            finalizeTimeoutTask?.cancel()
            finalizeTimeoutTask = nil
            AppLog.network.info("final asr=\(asr ?? 0, privacy: .public) llm=\(llm ?? 0, privacy: .public)")
            onFinal?(text)
        case .warning(let code, let message):
            AppLog.network.warning("服务端 warning [\(code, privacy: .public)]: \(message, privacy: .public)")
            onWarning?(message)
        case .error(_, let message):
            finalizeTimeoutTask?.cancel()
            finalizeTimeoutTask = nil
            AppLog.network.error("服务端错误: \(message, privacy: .public)")
            onError?(message)
        case .unknown:
            break
        }
    }

    // MARK: - 内部工具

    private func sendJSON(_ payload: [String: Any]) {
        guard !closed, let task else { return }
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let raw = String(data: data, encoding: .utf8) else { return }
        task.send(.string(raw)) { error in
            if let error {
                AppLog.network.error("控制帧发送失败: \(error.localizedDescription, privacy: .public)")
            }
        }
    }
}
